File: Skiena/binarySearch.java
package Skiena;

public class binarySearch {
	int array[];

	binarySearch(int a[]) {
		this.array = new int[a.length];

		// Deep Copy
		for (int i = 0; i < a.length; i++) {
			this.array[i] = a[i];
		}
	}

	public int binarySearchX(int x, int left, int right) {
		while (left <= right) {
			int mid = (left + right) / 2;
			if (array[mid] == x)
				return mid;
			else if (array[mid] > x)
				right = mid - 1;
			else
				left = mid + 1;
		}

		return -1;
	}
}

File: LeetCode/src/BinarySearch.py
def split_search(list: list, left, right, search):
    if left <= right:
        mid = (left + right) // 2
        if list[mid] == search:
            return mid
        elif list[mid] < search:
            return split_search(list, mid + 1, right, search)
        elif list[mid] > search:
            return split_search(list, left, mid - 1, search)
    return -1


def split_insert_search(list: list, left, right, search):
    mid = (left + right) // 2
    if left < right:
        mid = (left + right) // 2
        if list[mid] == search:
            return mid
        elif list[mid] < search:
            return split_insert_search(list, mid + 1, right, search)
        elif list[mid] > search:
            return split_insert_search(list, left, mid, search)
    return mid


# https://leetcode.com/problems/search-insert-position/
def searchInsert(nums, target: int) -> int:
    return split_insert_search(nums, 0, len(nums), target)


def binary_search(l: list, search) -> int:
    return split_search(l, 0, len(l) - 1, search)
